flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io;
+
+import org.apache.flink.annotation.Internal;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for listening on data availability using
+ * {@link CompletableFuture}. For usage check out for example {@link AsyncDataInput}.
+ */
+@Internal
+public interface AvailabilityListener {
+	/**
+	 * Constant that allows to avoid volatile checks {@link CompletableFuture#isDone()}. Check
+	 * {@link #isAvailable()} for more explanation.
+	 */
+	CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
+
+	/**
+	 * @return true if is finished and for example end of input was reached, false otherwise.
+	 */
+	boolean isFinished();
+
+	/**
+	 * Check if this instance is available for further processing.
+	 *
+	 * <p>When hot looping to avoid volatile access in {@link CompletableFuture#isDone()} user of
+	 * this method should do the following check:
+	 * <pre>
+	 * {@code
+	 *	AvailabilityListener input = ...;
+	 *	if (input.isAvailable() == AvailabilityListener.AVAILABLE || input.isAvailable().isDone()) {
+	 *		// do something;
+	 *	}
+	 * }
+	 * </pre>
+	 *
+	 * @return a future that is completed if there are more records available. If there are more
+	 * records available immediately, {@link #AVAILABLE} should be returned. Previously returned
+	 * not completed futures should become completed once there is more input available or if
+	 * the input {@link #isFinished()}.
+	 */
+	CompletableFuture<?> isAvailable();
```

Comment:
Maybe you can simplify the contract of this interface, by having only one method that returns 3 possible states of an underlying data stream? In nutshell, implementations of this interface have three situations:
 * stream has finished;
 * stream has not finished, and there is an immediate data ready to be processed;
 * stream has not finished, but the caller needs to wait when data is ready.

One way, you can apply the same trick as with `CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);`.
Basically, having pre-defined different constant dummy `CompletableFuture` that would be marker of `isFinished`:

```
public interface AvailabilityListener {
    CompletableFuture<?> FINISHED = CompletableFuture.completedFuture(null);
    CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);

    CompletableFuture<?> getStatusFuture();
}
```

Reply:
At first I liked the idea, however:
1. I think it is actually more difficult for the class user. Having `isFinished()` method is pretty self explanatory, while you would have to read about `FINISHED` constant in the java doc. Yes, it solves the connection between `isFinished()` and `isAvailable()`, but the issue that I mentioned IMO kind of balances this out
2. It would be actually quite a bit harder to implement and could cause new potential issues/bugs, like someone switched `getStatusFuture()` to `FINISHED` and then accidentally back to `AVAILABLE`.

I think there is another solution, with exposing to the implementator some context class similar to the one used in mailbox, while keeping the outer interface named differently.

However same comment as I wrote to Stefan, I think this will have to be kept in sync with `SourceReader` and we should discuss it there.

Reply:
There might be even another way that we can discuss here. IIRC one core problem in this PR with `CompletableFuture` is that it is a class and not an interface, so the implementation that is already completed cannot be more effcient. Did we ever consider returning a `CompletionStage` instead, which could be backed by a completable future in the cases that require waiting, and with some efficient completed implementations? All return an enum that indicates either `AVAILABLE` or `FINISHED`.

Reply:
Something ala `CompletionStage<Status> inputStatus()`.

Reply:
No, we haven't thought about this. This also might be a good idea. I'll merge this as it is now, and as I wrote before, let's revisit it later after depending PRs on this are merged with context of propose `SourceReader` in mind.